pkg/lib/controller-runtime/source/dynamic.go

```diff
@@ -0,0 +1,196 @@
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/util/workqueue"
+	ctrlcache "sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var log = logf.Log.WithName("source")
+
+// Dynamic sources events for all kinds exposed in discovery.
+type Dynamic struct {
+	// Injected by the controller manager
+	client discovery.ServerResourcesInterface
+	cache  ctrlcache.Cache
+	stop   <-chan struct{}
+
+	mu      sync.RWMutex
+	sources map[schema.GroupVersionKind]source.Source
+}
+
+var (
+	_ source.Source    = &Dynamic{}
+	_ inject.Cache     = &Dynamic{}
+	_ inject.Stoppable = &Dynamic{}
+)
+
+// Start is internal and should be called only by the Controller to register an EventHandler with the Informer to enqueue reconcile.Requests.
+func (d *Dynamic) Start(handle handler.EventHandler, queue workqueue.RateLimitingInterface, predicates ...predicate.Predicate) error {
+	if d.client == nil {
+		return fmt.Errorf("must inject rest config before starting dynamic source")
+	}
+	if d.cache == nil {
+		return fmt.Errorf("must inject cache before starting dynamic source")
+	}
+	if d.stop == nil {
+		return fmt.Errorf("must inject stop channel before starting dynamic source")
+	}
+
+	// TODO(njhale): Stop sources when GVKs are removed
+	var (
+		synced sync.WaitGroup
+		once   sync.Once
+	)
+	synced.Add(1)
+
+	go wait.Until(func() {
+		defer once.Do(func() {
+			synced.Done()
+		})
+
+		informable, err := d.InformableGVKs()
+		if err != nil {
+			log.Error(err, "could not get available gvks")
+		}
+
+		for _, gvk := range informable {
+			if d.sourceStarted(gvk) {
+				log.V(4).Info("dynamic source already started", "gvk", gvk)
+				continue
+			}
+
+			if err := d.startSource(gvk, handle, queue, predicates...); err != nil {
+				log.Error(err, "failed to start dynamic source", "gvk", gvk)
+			}
+		}
+
+	}, 15*time.Second, d.stop)
+
+	synced.Wait()
+
+	return nil
+}
+
+var (
+	knownInformables []schema.GroupVersionKind
+	informablesLock  sync.RWMutex
+)
+
+// InformableGVKs returns the set of GVKs available in discovery that support the list and watch verbs (informable).
+func (d *Dynamic) InformableGVKs() (informables []schema.GroupVersionKind, err error) {
+	if d.client == nil {
+		return nil, fmt.Errorf("no client")
+	}
+
+	var resourceLists []*metav1.APIResourceList
+	resourceLists, err = d.client.ServerPreferredResources()
+	if err != nil {
+		informablesLock.RLock()
+		defer informablesLock.RUnlock()
+		log.Error(err, "could not get preferred resources, returning latest known gvks")
+		return knownInformables, nil
```

Comment:
Going to memoize the last successful `APIResourceList` and whittle it down with direct checks against the discovery client instead. This will let us weed out the problem resource types before returning.